enhancements/cluster-logging/performance-tuning.md

```diff
@@ -0,0 +1,241 @@
+
+### Non-Goals
+
+- Not offering guaruanteed reliable end-to-end delivery (limited by reliability of source and sink)
+- Not exposing underlying vector configuration details.
+- No rate limiting - already provided by `outputs[].limit` field.
+- No plans to support fluentd.
+
+## Proposal
+
+### Workflow Description
+
+**logging administrator** is a human user responsible for setting output tuning parameters as needed.
+
+### API Extensions
+
+New field `ClusterLogForwarder.output[].tuning` with the following sub-fields:
+
+#### delivery: Fast|Safe
+
+`Fast`: Minimize resource use (CPU, disk, network) and maximize throughput in normal operation. Use when:
+- It is acceptable to lose some logs due to restarts, failures or heavy load.
+- The logging system will run close to it's limits, and may sometimes be overloaded.
+- It is important to minimize memory, CPU, disk and network costs.
+
+`Safe`: Minimize log loss due to restarts or failures. Use when:
+- It is important to avoid log loss due to restarts or failures.
+- The logging system is not expected to be pushed close to its limits.
+- Resources (memory, CPU, disk) are sufficient for log throughput to exceed expected peak load.
+
+**Note**: The log collector _cannot_ guarantee reliable end-to-end delivery.
+It has no control over the reliability or throughput of sources and sinks.
+The `delivery` policy tunes the log collector's "best effort".
+
+#### compression: string
+
+Enable compression if supported
+- Value indicates compression type (e.g. "zlib", "gzip"), valid values depend on the output type.
+- Error if output does not support compression or does not recognize the value.
+
+#### batch: object
+A "batch" is the content of a single "write" or "send" operation.
+Defaults are determined by the underlying sink.
+
+- `maxRecords`: Max number of records to include in a batch.
+- `maxBytes`: Max size in bytes of a batch
+
+#### reconnect: object
+Controls connect and reconnect attempts.
+Defaults are determined by the underlying sink.
+
+- `minDelay`: minimum delay between attempts.
+- `maxDelay`: maximum delay between attempts.
+
+`minDelay` for the first attempt, delay increases up to `maxDelay` and then repeats.
+Back-off algorithm is determined by the underlying sink.
+
+### Implementation Details
+
+#### Causes of log loss
+
+1. *Overload*: Logs are produced faster than they can be processed.
+2. *Faults*: Collector restarts, network errors, remote store problems etc.
+
+Log loss can be avoided for _temporary_ faults or overloads.
+Persistent buffers and/or acknowledgements can store or re-send lost in-memory data.
+
+_Sustained overload_ lasting long enough to exceed buffering capacity _will_ cause data loss.
+The only remedy is to ensure that the log collector and store can keep up with the rate of log production.
+
+#### Acknowledgements
+
+[End-to-end acknowledgement](https://vector.dev/docs/about/under-the-hood/architecture/end-to-end-acknowledgements) 
```

Comment:
I think we need an explicit call out that we intend to implement via e2e ack instead of file buffering.  This describes e2e but does not really say that is how it is being implemented.  We otherwise risk the repeated question and inquiries that there is no buffering enabled.  Additionally, we need to test and verify that e2e acks=true is equivalent to e2e acks=false + buffering

Reply:
Will have a think about how to phrase this, perhaps more explicit naming around acks vs. persistent buffers will solve the fast/safe naming problem.

Reply:
I would also suggest that rather than automatically enabling e2e, users should be able to choose whether to enable persistent buffering or let e2e do it for them.
This would also improve supportability in case e2e does not work.
When e2e does not work, the user can enable persistent buffering to achieve the purpose of delivery: slow.

Reply:
@kattz-kawa agreed, my current thinking is to 
- enable e2e without buffering _by default_ - because it provides better reliability without much resource penalty
- allow manually forcing internal buffering as a user option, for special situations where e2e is inadequate
-  allow turning off e2e *without* buffering for special situations where reliability is not important and even the low cost of e2e is not wanted

Reply:
See new draft: uses AtMostOnce, AtLeastOnce and AtLeastOncePersistent which I think is both clearer and  easier to explain. @jcantrill we can add AtLeastOncePersistent to your v2 API work or leave it for a later release.
It provides an escape hatch if e2e doesn't work in some situations, the implementation is just a combination of  e2e+persistent so probably not hard to add.